src/main/java/org/xbill/DNS/DNSSession.java

```diff
@@ -0,0 +1,129 @@
+package org.xbill.DNS;
+
+import static java.lang.String.format;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.xbill.DNS.exceptions.AdditionalDetail;
+import org.xbill.DNS.exceptions.InvalidZoneDataException;
+import org.xbill.DNS.exceptions.LookupFailedException;
+import org.xbill.DNS.exceptions.RedirectOverflowException;
+import org.xbill.DNS.exceptions.ServerFailedException;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+public class DNSSession {
```

Comment:
- Session somewhat implies it needs closing or that it expires, not sure this really applies here
- The async stuff creates some interesting threading issues once there are more config options than a simple integer. E.g. what if there's a cache setter, and the cache is changed while an async lookup is still in progress?
- The existing Cache class is not thread safe (IIRC), so it cannot be reused

Reply:
I like immutable classes, with the drawback that configuration typically would be through a long set of constructor args or via the Builder pattern which is a bit verbose. Let me think some more about this

Reply:
A builder seems appropriate - and Lombok to the rescue ;-)